File: go/nvagentrt/guardrails/guardrails.go
// SPDX-FileCopyrightText: Copyright (c) 2026, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
// SPDX-License-Identifier: Apache-2.0

// Package guardrails provides shorthand access to NVAgentRT guardrail registration.
//
// Guardrails are priority-ordered middleware that sanitize or gate tool and LLM
// calls. They run in priority order (lower values first). Function names drop
// the "Guardrail" suffix found in the parent nvagentrt package.
//
// Three guardrail categories are supported for both tools and LLMs:
//   - SanitizeRequest: modifies outgoing request arguments/parameters.
//   - SanitizeResponse: modifies incoming response data.
//   - ConditionalExecution: gates whether the call should proceed at all.
//
// Example usage:
//
//	import "github.com/nvidia/nvagentrt/go/nvagentrt/guardrails"
//
//	// Register a tool request sanitizer that redacts sensitive fields.
//	err := guardrails.RegisterToolSanitizeRequest("redact-pii", 10,
//	    func(name string, args json.RawMessage) json.RawMessage {
//	        // ... redact PII from args ...
//	        return args
//	    },
//	)
//
//	// Later, remove it.
//	_ = guardrails.DeregisterToolSanitizeRequest("redact-pii")
package guardrails

import (
	"encoding/json"

	"github.com/nvidia/nvagentrt/go/nvagentrt"
)

// --- Tool Sanitize Request ---

// RegisterToolSanitizeRequest registers a guardrail that sanitizes tool request
// arguments before they are passed to the tool. The callback receives the tool
// name and arguments JSON and must return the (possibly modified) arguments.
// Guardrails run in priority order (lower values first). This is a shorthand
// for [nvagentrt.RegisterToolSanitizeRequestGuardrail].
func RegisterToolSanitizeRequest(name string, priority int32, fn nvagentrt.ToolSanitizeFunc) error {
	return nvagentrt.RegisterToolSanitizeRequestGuardrail(name, priority, fn)
}

// DeregisterToolSanitizeRequest removes a tool sanitize-request guardrail by
// name. This is a shorthand for [nvagentrt.DeregisterToolSanitizeRequestGuardrail].
func DeregisterToolSanitizeRequest(name string) error {
	return nvagentrt.DeregisterToolSanitizeRequestGuardrail(name)
}

// --- Tool Sanitize Response ---

// RegisterToolSanitizeResponse registers a guardrail that sanitizes tool
// response data before it is returned to the caller. The callback receives the
// tool name and response JSON and must return the (possibly modified) response.
// This is a shorthand for [nvagentrt.RegisterToolSanitizeResponseGuardrail].
func RegisterToolSanitizeResponse(name string, priority int32, fn nvagentrt.ToolSanitizeFunc) error {
	return nvagentrt.RegisterToolSanitizeResponseGuardrail(name, priority, fn)
}

// DeregisterToolSanitizeResponse removes a tool sanitize-response guardrail by
// name. This is a shorthand for [nvagentrt.DeregisterToolSanitizeResponseGuardrail].
func DeregisterToolSanitizeResponse(name string) error {
	return nvagentrt.DeregisterToolSanitizeResponseGuardrail(name)
}

// --- Tool Conditional Execution ---

// RegisterToolConditionalExecution registers a guardrail that conditionally
// gates tool execution. The callback returns nil to allow execution or a
// non-nil pointer to an error message string to reject it. This is a shorthand
// for [nvagentrt.RegisterToolConditionalExecutionGuardrail].
func RegisterToolConditionalExecution(name string, priority int32, fn nvagentrt.ToolConditionalFunc) error {
	return nvagentrt.RegisterToolConditionalExecutionGuardrail(name, priority, fn)
}

// DeregisterToolConditionalExecution removes a tool conditional-execution
// guardrail by name. This is a shorthand for
// [nvagentrt.DeregisterToolConditionalExecutionGuardrail].
func DeregisterToolConditionalExecution(name string) error {
	return nvagentrt.DeregisterToolConditionalExecutionGuardrail(name)
}

// --- LLM Sanitize Request ---

// RegisterLlmSanitizeRequest registers a guardrail that sanitizes LLM request
// parameters (HTTP method, URL, headers, body) before the call is made. This is
// a shorthand for [nvagentrt.RegisterLlmSanitizeRequestGuardrail].
func RegisterLlmSanitizeRequest(name string, priority int32, fn nvagentrt.LLMRequestFunc) error {
	return nvagentrt.RegisterLlmSanitizeRequestGuardrail(name, priority, fn)
}

// DeregisterLlmSanitizeRequest removes an LLM sanitize-request guardrail by
// name. This is a shorthand for [nvagentrt.DeregisterLlmSanitizeRequestGuardrail].
func DeregisterLlmSanitizeRequest(name string) error {
	return nvagentrt.DeregisterLlmSanitizeRequestGuardrail(name)
}

// --- LLM Sanitize Response ---

// RegisterLlmSanitizeResponse registers a guardrail that sanitizes LLM response
// JSON before it is returned to the caller. This is a shorthand for
// [nvagentrt.RegisterLlmSanitizeResponseGuardrail].
func RegisterLlmSanitizeResponse(name string, priority int32, fn nvagentrt.JSONFunc) error {
	return nvagentrt.RegisterLlmSanitizeResponseGuardrail(name, priority, fn)
}

// DeregisterLlmSanitizeResponse removes an LLM sanitize-response guardrail by
// name. This is a shorthand for [nvagentrt.DeregisterLlmSanitizeResponseGuardrail].
func DeregisterLlmSanitizeResponse(name string) error {
	return nvagentrt.DeregisterLlmSanitizeResponseGuardrail(name)
}

// --- LLM Conditional Execution ---

// RegisterLlmConditionalExecution registers a guardrail that conditionally
// gates LLM execution. The callback receives LLM request parameters and returns
// nil to allow execution or a non-nil pointer to an error message string to
// reject it. This is a shorthand for
// [nvagentrt.RegisterLlmConditionalExecutionGuardrail].
func RegisterLlmConditionalExecution(name string, priority int32, fn nvagentrt.LLMConditionalFunc) error {
	return nvagentrt.RegisterLlmConditionalExecutionGuardrail(name, priority, fn)
}

// DeregisterLlmConditionalExecution removes an LLM conditional-execution
// guardrail by name. This is a shorthand for
// [nvagentrt.DeregisterLlmConditionalExecutionGuardrail].
func DeregisterLlmConditionalExecution(name string) error {
	return nvagentrt.DeregisterLlmConditionalExecutionGuardrail(name)
}

// --- Tool Conditional Execution (standalone) ---

// ToolConditionalExecution runs the registered tool conditional execution
// guardrail chain. Returns nil if all pass, or an error if blocked. This is a
// shorthand for [nvagentrt.ToolConditionalExecution].
func ToolConditionalExecution(name string, args json.RawMessage) error {
	return nvagentrt.ToolConditionalExecution(name, args)
}

// --- LLM Conditional Execution (standalone) ---

// LlmConditionalExecution runs the registered LLM conditional execution
// guardrail chain. Returns nil if all pass, or an error if blocked. This is a
// shorthand for [nvagentrt.LlmConditionalExecution].
func LlmConditionalExecution(request json.RawMessage) error {
	return nvagentrt.LlmConditionalExecution(request)
}

File: python/nvagentrt/guardrails.py
# SPDX-FileCopyrightText: Copyright (c) 2026, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: Apache-2.0

"""Guardrail registration for tools and LLMs.

Guardrails run inside the middleware pipeline and can sanitize or gate requests
and responses. They are priority-ordered (ascending) and registered by name.

**Tool guardrails** — callback signatures:

    register_tool_sanitize_request(name, priority, fn)
        ``fn(tool_name: str, args: Any) -> Any`` — sanitize tool arguments.

    register_tool_sanitize_response(name, priority, fn)
        ``fn(tool_name: str, result: Any) -> Any`` — sanitize tool result.

    register_tool_conditional_execution(name, priority, fn)
        ``fn(tool_name: str, args: Any) -> Optional[str]`` — return ``None``
        to allow, or a rejection reason to block.

**LLM guardrails** — callback signatures:

    register_llm_sanitize_request(name, priority, fn)
        ``fn(request: LLMRequest) -> LLMRequest`` — sanitize the LLM request.

    register_llm_sanitize_response(name, priority, fn)
        ``fn(response: LLMResponse) -> LLMResponse`` — sanitize the LLM response.

    register_llm_conditional_execution(name, priority, fn)
        ``fn(request: LLMRequest) -> Optional[str]`` — return ``None``
        to allow, or a rejection reason to block.

Each ``register_*`` has a corresponding ``deregister_*`` that takes the name
and returns ``True`` if a guardrail was found and removed.

Example::

    import nvagentrt

    def redact_pii(tool_name, args):
        # Sanitize PII from tool arguments
        return {k: "***" if "ssn" in k else v for k, v in args.items()}

    nvagentrt.guardrails.register_tool_sanitize_request("pii-filter", 10, redact_pii)
"""

from nvagentrt._native import (
    nvagentrt_deregister_llm_conditional_execution_guardrail as deregister_llm_conditional_execution,
)
from nvagentrt._native import (
    nvagentrt_deregister_llm_sanitize_request_guardrail as deregister_llm_sanitize_request,
)
from nvagentrt._native import (
    nvagentrt_deregister_llm_sanitize_response_guardrail as deregister_llm_sanitize_response,
)
from nvagentrt._native import (
    nvagentrt_deregister_tool_conditional_execution_guardrail as deregister_tool_conditional_execution,
)
from nvagentrt._native import (
    nvagentrt_deregister_tool_sanitize_request_guardrail as deregister_tool_sanitize_request,
)
from nvagentrt._native import (
    nvagentrt_deregister_tool_sanitize_response_guardrail as deregister_tool_sanitize_response,
)
from nvagentrt._native import (
    nvagentrt_register_llm_conditional_execution_guardrail as register_llm_conditional_execution,
)
from nvagentrt._native import (
    # LLM guardrails
    nvagentrt_register_llm_sanitize_request_guardrail as register_llm_sanitize_request,
)
from nvagentrt._native import (
    nvagentrt_register_llm_sanitize_response_guardrail as register_llm_sanitize_response,
)
from nvagentrt._native import (
    nvagentrt_register_tool_conditional_execution_guardrail as register_tool_conditional_execution,
)
from nvagentrt._native import (
    # Tool guardrails
    nvagentrt_register_tool_sanitize_request_guardrail as register_tool_sanitize_request,
)
from nvagentrt._native import (
    nvagentrt_register_tool_sanitize_response_guardrail as register_tool_sanitize_response,
)

__all__ = [
    "register_tool_sanitize_request",
    "deregister_tool_sanitize_request",
    "register_tool_sanitize_response",
    "deregister_tool_sanitize_response",
    "register_tool_conditional_execution",
    "deregister_tool_conditional_execution",
    "register_llm_sanitize_request",
    "deregister_llm_sanitize_request",
    "register_llm_sanitize_response",
    "deregister_llm_sanitize_response",
    "register_llm_conditional_execution",
    "deregister_llm_conditional_execution",
]
